### PR TITLE
Navigation confirmation for leaving project page without saving description

### PR DESCRIPTION
**Pull request by stowellm**
Addresses #1828 
Even added some tests to cover editing project descriptions and ensuring the pop-up I implemented here works as intended (that is, asking for navigation confirmation if the project description was modified, not saved/canceled, and the user tried leaving the page).

### CODE REVIEW

**Comment by stowellm** [comment]
@kcarcia I found some warning that was popping up in the console if the description was not set when viewing a project as a non-owner.  I don't think that was causing the issue, but nevertheless, I fixed that.

This may sound bizarre, but this is working for me again.  Next time you have a moment and we're both in the lab, can we look at this on your machine again?

**Comment by stowellm** [comment]
@kcarcia Thanks for helping me find the issue :smile:.  **Now** I suspect it'll work.  Ready for review pending Travis pass.

**Comment by kcarcia** [comment]
@stowellm I hate to be a nudge, but is it possible to implement this wherever summernote exists? Ideally, it'd to be great to make this an extension of summernote itself.

My fear is that  if we create an expectation that our editors prevent you from losing your work but don't follow through with that in all places, the users may be sad.

**Comment by stowellm** [comment]
@kcarcia This is now implemented anywhere summernote exists!  @jaypoulz asked to take this review, so I'm assigning to him (and assuming he'll assign to you if it passes your review).

The places to note that have summernote where this works is:
- project show page
- visualization page
- user bio page

File: app/assets/javascripts/shared/confirm_nav.js.coffee [hunk 1]
```diff
@@ -0,0 +1,42 @@
+setupNavConfirmation = () ->
+  # Initialize a variable that determines whether or not to ask
+  # for confirmation to leave the page, and keep track of the
+  # current description content
+  confirm_nav = false
+  if $('.summernote').code()
+    desc = $('.summernote').code().trim()
+  else
```
Comment (jaypoulz) [inline app/assets/javascripts/shared/confirm_nav.js.coffee]: This 'else' is useless:
```
desc = ''
if $('.summernote').code()
    desc = $('.summernote').code().trim()
```

**Comment by jaypoulz** [comment]
@stowellm let's go through this review together next time you're in the lab. there are a few things that confuse me, and I think it'll be easier this way.

File: app/assets/javascripts/shared/confirm_nav.js.coffee [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+    desc = ''
+
+  # Disable the navigation pop-up if the user is saving or canceling
+  if $('#content-save-btn')?
```
Comment (jaypoulz) [inline app/assets/javascripts/shared/confirm_nav.js.coffee]: if can go, add return

File: app/assets/javascripts/shared/confirm_nav.js.coffee [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+      confirm_nav = true
+  if $('#add-content-image')?
+    $('#add-content-image').click ->
+      confirm_nav = true
+
+  # Pop up a warning if the description has changed
+  $(window).on 'beforeunload', ->
+    if $('.summernote').code()
+      desc_new = $('.summernote').code().trim()
+    else
```
Comment (jaypoulz) [inline app/assets/javascripts/shared/confirm_nav.js.coffee]: ditto

File: test/integration/edit_proj_desc_test.rb [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+  setup do
+    @project = projects(:description_test)
+    Capybara.current_driver = :webkit
+    Capybara.default_wait_time = 15
+  end
+
+  teardown do
+    finish
+  end
+
```
Comment (jaypoulz) [inline test/integration/edit_proj_desc_test.rb]: def function
end

**Comment by kcarcia** [comment]
@stowellm great work!

File: test/integration/edit_proj_desc_test.rb (hunk 4 above)
Comment (stowellm) [inline test/integration/edit_proj_desc_test.rb]: Function has been def'ed.  Diff not suppressed because was not addressed at this very line.

**Comment by stowellm** [comment]
@jaypoulz ready for re-review

File: app/assets/javascripts/shared/confirm_nav.js.coffee [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+    confirm_nav = true
+
+  # Pop up a warning if the description has changed
+  $(window).on 'beforeunload', ->
+    desc_new = ''
+    if $('.summernote').code()
+      desc_new = $('.summernote').code().trim()
+
+    if (confirm_nav && (desc_new != desc))
+      return "You have attempted to leave this page. If you have made any " +
```
Comment (jaypoulz) [inline app/assets/javascripts/shared/confirm_nav.js.coffee]: Now that I've run the code, I'm confused as to what this is for.

My page reads:
"This page is asking you to confirm that you want to leave - data you have entered may not be saved."

Comment (stowellm) [inline app/assets/javascripts/shared/confirm_nav.js.coffee]: Wat

Comment (jaypoulz) [inline app/assets/javascripts/shared/confirm_nav.js.coffee]: Firefox doesn't wanna deal with yo nonsense. GG

Comment (stowellm) [inline app/assets/javascripts/shared/confirm_nav.js.coffee]: For other readers: Jeremy and I exchanged some Hangouts messages/screenshots that shows that Chrome displays the message I wrote while Firefox displays its own navigation message.  Either way, you get an appropriate navigation confirmation dialog, so we're leaving this as is.

**Pull request merged**